src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -110,6 +110,30 @@ public async ValueTask<IProjectContextInfo> GetProjectAsync(string projectId, Ca
             return packageReferences.SelectMany(e => e).Select(pr => PackageReferenceContextInfo.Create(pr)).ToArray();
         }
 
+        public async ValueTask<IReadOnlyCollection<IPackageReferenceContextInfo>> GetTransitivePackagesAsync(
+            IReadOnlyCollection<string> projectIds,
+            CancellationToken cancellationToken)
+        {
+            Assumes.NotNullOrEmpty(projectIds);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            IReadOnlyList<NuGetProject> projects = await GetProjectsAsync(projectIds, cancellationToken);
+
+            // Read transitive package references from package reference style projects.
+            IEnumerable<Task<IEnumerable<PackageReference>>> tasks = projects
```

Comment:
From a readability perspective I'm not a fan of long linq statements. 
Can the method maybe be added to a shared PackageReference type? 
Fewer casts, better readability.

Reply:
My pseudo code suggestion:

```cs
List<Task<IReadOnlyCollection<PackageReference>>> tasks = new List<....>();
foreach (var project in projects)
{
  if (project is LegackPackageReferenceProject legacyPRProject)
  {
    tasks.Add(legacyPRProject.GetTransitivePackagesAsync(cancellationToken));
  }
  else if (project is NetCorePackageReferenceProject cpsProject)
  {
    tasks.Add(cpsProject.GetTransitivePackagesAsync(cancellationToken));
  }
}
IReadOnlyCollection<PackageReference>[] transitivePackageReferencesArray = await Task.WhenAll(tasks, cancellationToken);

```

I think this might be fewer lines, despite all the curly brace lines. The LINQ expression could be shorter, and more performant, by using `.OfKind<T>()` instead of `.Where(p => p is T).Select(p => (T)p)`. But in my opinion LINQ does not always make code more readable, especially if you're enumerating the collection twice and then concating them into a single result.

Reply:
I modified the code using your suggestion.